Kb driven input rework

BurningTreeC opened the pull request:
This PR reworks the keyboard-driven-input macros to the new Syntax using procedures and conditionals.

BurningTreeC commented:
There is some code duplication now because Procedures won't let me use `$variables$` when called as parameters but in the long term I believe this pays off

Jermolene commented:
Thanks @BurningTreeC. I notice that there are still quite a few `\defines`, are they needed?

Are there any backwards compatibility impacts?

BurningTreeC commented:
Hello @Jermolene , the `\defines` are in the macros calling the `change-input-tab-next` and `change-input-tab-previous` procedures. In terms of backwards compatibility the `change-input-tab` macro is no more, that's why I had to change the tiddlers that called it. In those tiddlers I left the `\defines` stay because it would be a bigger rework needed to change everything to procedures.

saqimtiaz commented on core/ui/AdvancedSearch/Filter.tid:
```diff
@@ -3,11 +3,19 @@ tags: $:/tags/AdvancedSearch
 \define lingo-base() $:/language/Search/
-\define set-next-input-tab(beforeafter:"after")
-<$macrocall $name="change-input-tab"
+\define set-next-input-tab()
```
macros should be replaced with procedures

BurningTreeC replied:
done

saqimtiaz commented on core/ui/AdvancedSearch/Filter.tid:
```diff
@@ -3,11 +3,19 @@ tags: $:/tags/AdvancedSearch
-\define set-next-input-tab(beforeafter:"after")
-<$macrocall $name="change-input-tab"
+\define set-next-input-tab()
+<$macrocall $name="change-input-tab-after"
```
$macrocall is deprecated in favour of $transclude.

BurningTreeC replied:
done

BurningTreeC commented on core/wiki/macros/keyboard-driven-input.tid:
```diff
@@ -1,105 +1,146 @@
-</$list>
+<$set name="tabsList" filter="[all[shadows+tiddlers]tag<tag>!has[draft.of]]">
+	<$let
+		currentState={{{ [<stateTitle>!is[missing]get[text]] ~[<defaultState>] }}}
```
done

pmario commented:
@BurningTreeC -- This PR seems to be not backwards compatible. -- or I am using it wrong whth "CTRL-ATL tab switching" in my AdvancedSearch Plugin.

The tag-picker macro seems to work with the new code.

pmario commented:
Did you change any macro-parameters?

BurningTreeC commented:
Hi @pmario - I had to change the `change-input-tab` footprint.
There are now two procedures `change-input-tab-after` and `change-input-tab-before` - that's because we cannot use `$afterorbefore$` in procedures.
Now, thinking about it - maybe I can build a function and thus make it compatible. I'll try

BurningTreeC commented:
Hi @pmario - now the `change-input-tab` procedure works like the old `change-input-tab` macro. That should solve your issues with the keyboard-driven-input procedures.

pmario commented:
I did a short test with the new version. -> It still does not work. ASP plugin works fine with v5.3.5 and the v5.3.6-prerelease

BurningTreeC commented:
Hi @pmario - I think this problem comes from the actions passed to the `change-input-tab` procedure not being substituted or how this is called. I used `$actions$` before in the macro. What's the new way of doing this in a procedure?

BurningTreeC commented:
@pmario I believe now I've found the problem:

ASP overrides the change-input-tab Macro with the tag $:/tags/Macro but I changed the tag to $:/tags/Global
Now I've incorporated your additions to the change-input-tab Macro here, so I believe there's no need to override it.
I also changed the tag back to `$:/tags/Macro` for now.
Could you test?

pmario commented:
>I also changed the tag back to $:/tags/Macro for now.
>Could you test?

Jeremy suggested that, if we change the syntax to v5.3.x we should add tags: `$:/tags/Macro` **and** `$:/tags/Global`

BurningTreeC commented:
Hi @pmario - can you test if it still works with both the `$:/tags/Macro` and the `$:/tags/Global` tag?

pmario commented:
That's strange. It only woks, if I also add both tags to my plugin. -- So for compatibility reasons $:/tags/Macro may be the way to go.

BurningTreeC commented:
@pmario - the `GlobalImportFilter` imports the Globals after the Macros and therefor "overwrites" the Macros:

```
[[$:/core/ui/PageMacros]] [all[shadows+tiddlers]tag[$:/tags/Macro]!is[draft]] [all[shadows+tiddlers]tag[$:/tags/Global]!is[draft]]
```

I believe it should be this way around:

```
[[$:/core/ui/PageMacros]] [all[shadows+tiddlers]tag[$:/tags/Global]!is[draft]] [all[shadows+tiddlers]tag[$:/tags/Macro]!is[draft]]
```

pmario commented:
I did create a new issue 

- #8683

BurningTreeC commented:
This PR is done @pmario 
Could you please test if your ASP works even if you delete your customized `change-input-tab` macro? (because it should)

pmario commented:
>This PR is done @pmario
Could you please test if your ASP works even if you delete your customized change-input-tab macro? (because it should)

@BurningTreeC -- It works without any changes to my plugin. Thx.

Jermolene commented:
Thanks @BurningTreeC
